### DISCUSSION TITLE
Add executeAsync

### DISCUSSION THREAD

**Discussion by anarkrypto** [next-safe-action/next-safe-action#137] [Ideas]
Tanstack react-query has a very useful method: **mutateAsync**

It allows us to get a promise which will resolve on success or throw on an error. This can for example be used to compose side effects.

For me it has been very useful in combination with **react-hook-form**. So I can pass my async mutation to the form's `handleSubmit` and use `isSubmitting`, `isSubmitted` / `isSubmitSuccessful` from `formState`.

However it is not possible with next-safe-action currently, we depend 100% of it's internal hook state.

To achieve the same effect of mutateAsync with next-safe-action execute, here is simple example I was working on:
```typescript
const executeAsync = React.useCallback(
    (input: InferIn<S>): Promise<void> => {
      setInput(input)
      setIsExecuting(true)

      return new Promise<void>((resolve, reject) => {
        startTransition(() => {
          return safeAction(input)
            .then(res => setResult(res ?? DEFAULT_RESULT))
            .catch(e => {
              if (isRedirectError(e) || isNotFoundError(e)) {
                throw e
              }

              reject(e)

              setResult({
                fetchError: isError(e) ? e.message : 'Something went wrong',
              })
            })
            .finally(() => {
              setIsExecuting(false)
              resolve()
            })
        })
      })
    },
    [safeAction],
  )
```

**Comment by TheEdoRan** [comment]
It would surely be useful, and it has already been discussed in:
- #72
- #94

My question is: does `startTransition()` work as expected when dealing with promises (being run inside a promise)? Even the latest `useActionState` hook from React returns the action function typed `void`, so I'm not sure if this design over React APIs will result in unwanted bugs that then users will experience when using this library.

**Reply by TheEdoRan** [reply]
Hey @anarkrypto, the addition of a `executeAsync` function in the library would be great. Have you found bugs when using your implementation of it?

**Reply by TheEdoRan** [reply]
Added `executeAsync` function to `useAction` and `useOptimisticAction` hooks return object in [v7.0.0-next.46](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.0.0-next.46). It seems to work just fine.

**Reply by anarkrypto** [reply]
> Added `executeAsync` function to `useAction` and `useOptimisticAction` hooks return object in [v7.0.0-next.46](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.0.0-next.46). It seems to work just fine.

Great @TheEdoRan!
I appreciate your work to make this happen!
Now this project has all the features I need in my projects to work with server actions.
 I am already using it in some projects. I will inform any possible issue if I have.
